src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs

```diff
@@ -0,0 +1,133 @@
+        "AreEquivalent",
+        "AreNotEquivalent",
+        "Contains",
+        "DoesNotContain",
+        "AllItemsAreNotNull",
+        "AllItemsAreUnique",
+        "IsSubsetOf",
+        "IsNotSubsetOf",
+        "AllItemsAreInstancesOfType",
+        "StartsWith",
+        "EndsWith",
+        "Matches",
+        "DoesNotMatch",
+    });
+
+    private static readonly LocalizableResourceString Title = new(nameof(Resources.AssertionArgsShouldAvoidConditionalAccessTitle), Resources.ResourceManager, typeof(Resources));
+    private static readonly LocalizableResourceString MessageFormat = new(nameof(Resources.AssertionArgsShouldAvoidConditionalAccessMessageFormat), Resources.ResourceManager, typeof(Resources));
+
+    internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+        DiagnosticIds.AssertionArgsShouldAvoidConditionalAccessRuleId,
+        Title,
+        MessageFormat,
+        description: null,
+        Category.Usage,
+        DiagnosticSeverity.Info,
+        isEnabledByDefault: true);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+        = ImmutableArray.Create(Rule);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationStartAction(context =>
+        {
+            if (context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftVisualStudioTestToolsUnitTestingAssert, out INamedTypeSymbol? assertSymbol))
+            {
+                context.RegisterOperationAction(context => AnalyzeOperation(context, assertSymbol), OperationKind.Invocation);
+            }
+
+            if (context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftVisualStudioTestToolsUnitTestingCollectionAssert, out INamedTypeSymbol? collectionAssertSymbol))
+            {
+                context.RegisterOperationAction(context => AnalyzeOperation(context, collectionAssertSymbol), OperationKind.Invocation);
+            }
+
+            if (context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftVisualStudioTestToolsUnitTestingStringAssert, out INamedTypeSymbol? stringAssertSymbol))
+            {
+                context.RegisterOperationAction(context => AnalyzeOperation(context, stringAssertSymbol), OperationKind.Invocation);
+            }
+        });
+    }
+
+    private static void AnalyzeOperation(OperationAnalysisContext context, INamedTypeSymbol assertSymbol)
+    {
+        var invocationOperation = (IInvocationOperation)context.Operation;
+
+        // This is not an invocation of the expected assertion methods.
+        if (!SupportedMethodNames.Contains(invocationOperation.TargetMethod.Name)
+            || !SymbolEqualityComparer.Default.Equals(assertSymbol, invocationOperation.TargetMethod.ContainingType)
```

Comment:
This is a bit fragile as it would allow to mix method name and types (e.g. `CollectionAssert` and `IsTrue`) but I guess that's probably fine for what we do here.

Reply:
Refactored it. Each assertion type checks for only its own methods